### PR TITLE
getValueForExpr: add support for arrays

### PR DESCRIPTION
**Pull request by gilbox**
# getValueForExpr: add support for arrays

Solves #10613 by adding support for array indexes for the `items` attribute of `amp-list`.

### CODE REVIEW

**Comment by robinpokorny** [comment]
I find this unnecesary complicated, it would be sufficient to use (maybe better formated and/or with variables):

```diff
- if (!isObject(value) ||
+ if (!(isObject(value) || isArray(value))||
            value[part] === undefined ||
            !hasOwnProperty(value, part)) {
      value = undefined;
      break;
    }
```

The flow is harder to follow with the added `continue`s. And `arrayHasKey` feels really hacky: `hasOwnProperty` should replace that magic in the first part of it and the second part needs no `typeof`.

File: src/json.js [hunk 1]
```diff
@@ -95,13 +95,18 @@ export function getValueForExpr(obj, expr) {
...
-            value[part] === undefined ||
-            !hasOwnProperty(value, part)) {
-      value = undefined;
-      break;
+    if (isArray(value) && arrayHasKey(value, part)) {
+      value = value[part];
+      continue;
     }
-    value = value[part];
+    if (isObject(value) &&
```
Comment (jridgewell) [inline src/json.js]: We could likely just remove this in favor of a truthy check:

```js
if (value && value[part] !== undefined && hasOwnProperty(value, part)) {
  //...
}

**Comment by cschubiner** [comment]
Exciting stuff, thanks @gilbox!

File: src/json.js (hunk 1 above)
Comment (gilbox) [inline src/json.js]: image

Comment (gilbox) [inline src/json.js]: great feed back y'all! I've added a commit that I hope addresses all of your comments

**Comment by jridgewell** [comment]
Apologies for letting this fall through the cracks.

**Pull request merged**